### PR TITLE
RLPT generate synthetic data

### LINKED ISSUE #10: generate synthetic data script in test runner
**User Story**

As a DIBBs engineer running record linkage performance tests, I need a way to generate N synthetic patient records for performance testing.

**Acceptance Criteria**

A new script is added to the runner service that is capable of using Synthea to generate N synthetic patient records.  N should be a parameter loaded into the environment at runtime.

### PR DESCRIPTION
**Pull request by ericbuckley**
# Pull Request

## Description
Use [Synthea](https://github.com/synthetichealth/synthea) a synthetic dataset for running performance tests.

## Related Issues
closes #10 

## Additional Notes
This was much simpler than I thought, only needed a couple of changes here to set up the parameters and run the synthea command.  My questions for the reviewers are, should we [parameterize any more options](https://github.com/synthetichealth/synthea/blob/master/src/main/resources/synthea.properties) for running performance tests?

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [ ] I have added or updated test cases to cover my changes, if applicable.
- [ ] I have minimized the number of reviewers to include only those essential for the review.
- [ ] I have notified teammates in the review thread to build awareness.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

**Comment by ericbuckley** [comment]
cc @derekadombek @cbrinson-rise8

File: dedupe/record_linkage_testing/compose.yml [hunk 1]
```diff
@@ -10,6 +10,11 @@ services:
       context: .
       dockerfile: Dockerfile.runner
     command: ["bash", "scripts/test.sh"]
+    environment:
+      - POPULATION_SIZE=100
```
Comment (alhayward) [inline dedupe/record_linkage_testing/compose.yml:14]: What is the motivation behind N=100?

File: dedupe/record_linkage_testing/scripts/test.sh [hunk 2]
```diff
@@ -66,3 +66,11 @@ EOF
...
 
 curl -X POST http://api:8080/link-record -d "${EXAMPLE_FHIR}" --header "Content-Type: application/json"
+
+# Run Synthea with parameters for seeding, the population size and the output directory.
+# Additionally disable generating hospital and practitioner FHIR resources, as we only
+# want patient data and limit the results to only alive patients.
+# For a full list of parameters, see the [configuration file](https://github.com/synthetichealth/synthea/blob/master/src/main/resources/synthea.properties).
+java -jar synthea.jar --exporter.hospital.fhir.export=false --exporter.practitioner.fhir.export=false \
+    --generate.only_alive_patients=true \
+    -p "${POPULATION_SIZE}" -s "${SEED}" -cs "${CLINICIAN_SEED}" -o "${SYNTHEA_OUTPUT_DIR}"
```
Comment (alhayward) [inline dedupe/record_linkage_testing/scripts/test.sh:76]: Do we have examples of the output records?

**Review by alhayward — COMMENTED** [review]
Thanks for this! Couple questions:
* What parameters did the DIBBs team use in their final synthetic dataset before gaining LAC production data? We should solicit their recommendations/feedback. (Curious also if we have access to any previous LAC data.)
* Timeboxing with the knowledge that production data may look significantly different, the question of whether we should include additional parameters when generating the synthetic patient population should be driven by: What is the aim of this performance testing? Some considerations:
    * Do we intend to vary any parameters beyond size? Are we interested in testing the scalability of the algorithm in any other characteristics of patient population (e.g., vary number of duplicates, vary population distributions of fields used for linking, etc.)
    * If we are focused on **load testing**, tuning population parameters may be less important, as long as the distributions remain representative of the test population as N increases.
    * By default, a synthetic dataset for MA will be generated. Should we specify LAC (and any other parameters the DIBBs team used) to replicate the synthetic data used previously?
    * Do we want to make use of the `exporter.split_records` or `exporter.split_records.duplicate_data` [properties](https://github.com/synthetichealth/synthea/blob/master/src/main/resources/synthea.properties#L12-L13), to simulate duplicate data?
    * Should we consider `generate.append_numbers_to_person_names = false`, since the DIBBs algorithm expects name strings with no digits?

Down to chat further!

File: dedupe/record_linkage_testing/compose.yml (hunk 1 above)
Comment (ericbuckley) [reply dedupe/record_linkage_testing/compose.yml:14]: There's really no significance to this right now, I just needed a default N to create the variable.  10 or even 1 would have been fine too.

Eventually when the script is finished, you'll likely override that from the command line when you run the performance tests, so something like 

```sh
docker compose run --env POPULATION_SIZE=1000 runner
```

File: dedupe/record_linkage_testing/scripts/test.sh (hunk 2 above)
Comment (ericbuckley) [reply dedupe/record_linkage_testing/scripts/test.sh:76]: Synthea has a sample repo with some zip files that you can download to take a look.  Try this one https://github.com/synthetichealth/synthea-sample-data/blob/main/downloads/latest/synthea_sample_data_fhir_latest.zip

**Comment by ericbuckley** [comment]
@alhayward apologies for the delay, I missed this comment the other day.

> Thanks for this! Couple questions:
> 
>     * What parameters did the DIBBs team use in their final synthetic dataset before gaining LAC production data? We should solicit their recommendations/feedback. (Curious also if we have access to any previous LAC data.)

I'm not :100: sure on that, but I did use the same parameters that Marcelle and Daniel setup in their [Synthea generate and load script](https://github.com/CDCgov/phdi/blob/dan/generate-and-load-synthea-data/utils/generate_and_load_synthea/generate_and_load_synthea.py).  I'm assuming they used parameters similar to this, if not exactly the same, but I agree it be worth asking them to confirm in this PR (will do shortly).

> 
>     * Timeboxing with the knowledge that production data may look significantly different, the question of whether we should include additional parameters when generating the synthetic patient population should be driven by: What is the aim of this performance testing? Some considerations:
>       
>       * Do we intend to vary any parameters beyond size? Are we interested in testing the scalability of the algorithm in any other characteristics of patient population (e.g., vary number of duplicates, vary population distributions of fields used for linking, etc.)?
>       * If we are focused on **load testing**, tuning population parameters may be less important, as long as the distributions remain representative of the test population as N increases.

My initial intentions were to focus on database scalability, specifically the relationship of API latency (i.e. how long the client waits for a response from the API) to the number of Patient records.

>       * By default, a synthetic dataset for MA will be generated. Should we specify LAC (and any other parameters the DIBBs team used) to replicate the synthetic data used previously?

I think it makes sense to add an option for the location, so I'll make that a variable.  My initial research is telling me that there are only `--location` options for US States plus about 18 other countries.  Did you see any documentation that would allow us to specify a locale smaller than a State?

>       * Do we want to make use of the `exporter.split_records` or `exporter.split_records.duplicate_data` [properties](https://github.com/synthetichealth/synthea/blob/master/src/main/resources/synthea.properties#L12-L13), to simulate duplicate data?

I'm guessing the thought here is so that some of our API requests result in matches?  If so, I think including that option makes sense.

>       * Should we consider `generate.append_numbers_to_person_names = false`, since the DIBBs algorithm expects name strings with no digits?

I've run a few tests so far with that set to `true` and it seems to work fine.  Did you read something that indicates the algorithm won't work with numbers in names?

> 
> Down to chat further!

**Comment by ericbuckley** [comment]
> > ```
> >   * By default, a synthetic dataset for MA will be generated. Should we specify LAC (and any other parameters the DIBBs team used) to replicate the synthetic data used previously?
> > ```
> 
> I think it makes sense to add an option for the location, so I'll make that a variable. My initial research is telling me that there are only `--location` options for US States plus about 18 other countries. Did you see any documentation that would allow us to specify a locale smaller than a State?
> 

Ignore this, I see the path forward, it's not an option it's an argument!

**Comment by ericbuckley** [comment]
I'm going to close this PR.  As I started working on #11 and #12, a fair amount changed.  So I'm just going to combine #10, #11 & #12 into 1 PR.

**Pull request closed without merging**